### PR TITLE
ble service: Device Information Service

### PR DESCRIPTION
**Pull request by sdalu**
Main characteristics for the Device Information Service

### CODE REVIEW

**Comment by andrzej-kaczmarek** [comment]
Few comments:
- characteristics permissions should be configurable - for consistency, I'd make it the same way as configurable write permissions in GAP service so there's no need for separate enable and permissions syscfgs
- also for consistency I'd use helpers to set values instead of using extern struct
- there's no need for #error in case none of characteristics were enabled - this is explicitly allowed by spec
- default value for characteristics with no value set should be an empty string
- there are multiple consecutive empty lines in few places - use single empty line as separator

As an addition, it may be useful to have default value for some/each characteristics in syscfg (same as e.g. Device Name in GAP) so there's no need to set all of them in application.

**Comment by sdalu** [comment]
* I think "Model Number", "Manufacturer Name" will generally be know at compile time and won't change, so I don't feel comfortable at preallocating memory as for the Device Name in GAP. Is there a preprocessor trick to do conditional with string value?
* For the Firmware/Software we could use the image versioning, but I'm not sure which one makes more sense.

**Comment by jacobrosenthal** [comment]
@sdalu I took a shot at one of these a while ago and got busy and never PRed so thanks!

BTW I ended up finding newt's 'config' stuff, and filled a bunch of these from there
https://github.com/jacobrosenthal/mynewt-nimble-services/blob/master/services/dis/src/ble_svc_dis.c#L105

None of these are 'right', I just wanted to point out I think this is where a lot of these values are stored 'the newt way' right now for people who use newtmgr

File: net/nimble/host/services/dis/syscfg.yml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+# specific language governing permissions and limitations
+# under the License.
+#
+
+# Package: ble/host/services/dis
+
+syscfg.defs:
+    BLE_SVC_DIS_MODEL_NUMBER_READ_PERM:
+        description: >
+            Defines permissions for writing "Model Number" characteristics. Can
```
Comment (jacobrosenthal) [inline net/nimble/host/services/dis/syscfg.yml]: Same thing here, these should all be Defines permissions for *reading

File: net/nimble/host/services/dis/src/ble_svc_dis.c [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+        }, }
+    },
+
+    {
+        0, /* No more services. */
+    },
+};
+
+/**
+ * Simple write access callback for the device information service
```
Comment (jacobrosenthal) [inline net/nimble/host/services/dis/src/ble_svc_dis.c]: *read access callback

File: net/nimble/host/services/dis/syscfg.yml (hunk 1 above)
Comment (andrzej-kaczmarek) [reply net/nimble/host/services/dis/syscfg.yml]: Also looking at this now I wonder if we should have just one read_perm syscfg for all characteristics and enable for each characteristic separately? I am not sure if there is scenario for setting different permissions for each characteristic here.

Not really sure what is the proper solution here, just thinking out loud.

**Comment by andrzej-kaczmarek** [comment]
My proposal with default values was not to propagate each characteristic with "computed" value but simply have syscfg with simple string for all/some characteristics - just to e.g. define manufacturer or model name from app or settings easily. I think DIS implementation should be just generic one as it is now.

@jacobrosenthal I think reading these values from config and configuring DIS accordingly could be a nice example to include in one of sample apps - blehr or blecsc perhaps?

**Comment by jacobrosenthal** [comment]
@andrzej-kaczmarek Yeah either seems cool to me. We should figure out what some good newt defaults would be, that was my first pass a year ago.

@sdalu  No opinon on some defaults here, Technically theyre all optional in the spec so you could do it empty, so maybe ifdef guard this?

But currently if the user doesnt set any syscfg read permissions you get a compile error so well have to guard or default
```
Error: repos/apache-mynewt-core/net/nimble/host/services/dis/src/ble_svc_dis.c:101:1: error: 'ble_svc_dis_access' defined but not used [-Werror=unused-function]
 ble_svc_dis_access(uint16_t conn_handle, uint16_t attr_handle,
 ^
cc1: all warnings being treated as errors
```

File: net/nimble/host/services/dis/syscfg.yml (hunk 1 above)
Comment (sdalu) [reply net/nimble/host/services/dis/syscfg.yml]: Perhaps a `BLE_SVC_DIS_DEFAULT_READ_PERM`?
And propose the `MYNEWT_VAL_BLE_SVC_DIS_DEFAULT_READ_PERM` as a possible alternative for the `BLE_GATT_CHR_F_READ_*` values in `BLE_SVC_DIS_*_READ_PERM`

**Comment by sdalu** [comment]
See: https://github.com/apache/mynewt-nimble/pull/12

**Pull request closed without merging**